crates/dbsp/src/trace/spine_async/mod.rs

```diff
@@ -83,81 +76,360 @@ impl<B: Batch + Send + Sync> From<&Spine<B>> for CommittedSpine<B> {
     }
 }
 
-/// A unique identifier for batches we hold within levels.
-#[derive(SizeOf, Ord, Default, PartialOrd, Debug, Copy, Clone, Eq, PartialEq)]
-struct BatchIdent {
-    /// The number of nested batches, either 1 if no merging is going on or 2 if
-    /// this represents a pair of batches being merged. (Larger values would
-    /// represent multiway merges, which we currently don't do.)
-    #[size_of(skip)]
-    n: usize,
-    /// Which level this batch is from.
+/// A group of batches with similar sizes (as determined by [size_from_level]).
+struct Slot<B>
+where
+    B: Batch,
+{
+    /// Optionally, a pair of batches that are currently being merged.  These
+    /// batches are not in `loose_batches`.
     ///
-    /// We use this to find the two batches we want to remove when
-    /// we finished a merge.
-    level: usize,
-    /// A key that is unique within the level.
+    /// Invariant: the batches (if present) must be non-empty.
+    merging_batches: Option<[Arc<B>; 2]>,
```

Comment:
This looks like we can only process one merge per-level at any time? Is it a problem for small size classes?

Reply:
> This looks like we can only process one merge per-level at any time? 

Yes.

> Is it a problem for small size classes?

I don't think so. My theory is that merging has two benefits:
1. To make iteration and searching cheaper because we have fewer batches to iterate and search in our `CursorList`s.
2. In some cases only, to reduce the total amount of data, because weights summarize multiple items or cancel each other out or because filters drop data.

We only get these benefits when we complete a merge. An ongoing but incomplete merge only has a cost (the CPU we invested in it, plus memory or storage), with no benefits. Therefore, it is to our benefit to both complete as many merges as possible and to have as few ongoing merges as possible.

Since the smallest merges are cheapest, one might conclude that we should only do one merge at a time and that that should be the two smallest batches. In a static scenario, that would indeed be the right choice. But we tend to be getting a new smallest batch on every step. With that, the strategy of always doing the smallest merge piles up larger batches that never get merged. For example, suppose that we add a new batch with size 1 in every step, we start merging the smallest runs whenever we first a merge, and that a merge takes two steps. Then we end up with something like this, where each line is a step that adds a new batch of size 1, `()` designates that batches are being merged, and `->` shows that a merge was finished or started within the step:

```
1
1 1 -> (1 1)
1 (1 1)
1 1 (1 1) -> 1 1 2 -> 2 (1 1)
1 2 (1 1)
1 1 2 (1 1) -> 1 1 2 2 -> 2 2 (1 1)
1 2 2 (1 1)
1 1 2 2 (1 1) -> 1 1 2 2 2 -> 2 2 2 (1 1)
```

Basically we pile up runs that are slightly longer than the shortest and they never get merged.

So, we still want to complete as many merges as possible and to have as few ongoing merges as possible, but "as few ongoing merges as possible" needs to be more than one and needs to include merges that are bigger than the smallest possible merge. The new merge design does both. I don't know whether it is the best possible, it's only the one that I hit on first.

Reply:
The "fuel" assigned for merges at each step should be log N, where N is the total size of the set.
This is the amortized cost you will have to eventually pay.
So you shouldn't stop merging with the small batches of size 1 if there is anything else that can be merged.
But sometimes there won't be any candidate for merging yet.

Reply:
> The "fuel" assigned for merges at each step should be log N, where N is the total size of the set. This is the amortized cost you will have to eventually pay. So you shouldn't stop merging with the small batches of size 1 if there is anything else that can be merged. But sometimes there won't be any candidate for merging yet.

I don't think this answers anything though. We're not assigning fuel to steps, because merges and steps are asynchronous. Instead, the question is, at a given time, which merge or merges should we work on and how much? A merge of batches with length A and B has cost O(A+B) but the benefit when it completes is O(1) regardless of A and B (unless the merged version has length less than A+B).

Reply:
Yes, I am saying that if you are doing less than work log(N) you will have to pay for it later.
Ideally you always do around log(N) work in a merge.

Reply:
> Yes, I am saying that if you are doing less than work log(N) you will have to pay for it later. Ideally you always do around log(N) work in a merge.

Can you recommend any reading material? I don't know how to apply this or where it comes from.